gsuite_reports_rules/gsuite_drive_external_share.py

```diff
@@ -0,0 +1,130 @@
+        for p in acl_change_event["parameters"]
+    }
+
+    doc_title = parameters.get("doc_title", "TITLE_UNKNOWN")
+    old_visibility = parameters.get("old_visibility", "OLD_VISIBILITY_UNKNOWN")
+    new_visibility = parameters.get("visibility", "NEW_VISIBILITY_UNKNOWN")
+    target_user = parameters.get("target_user", "USER_UNKNOWN")
+    current_time = datetime.datetime.now()
+
+    if (new_visibility == "shared_externally" and
+            old_visibility == "private" and
+            not target_user.endswith(f"@{COMPANY_DOMAIN}")):
+        # This is a dangerous share, check exceptions:
+        for pattern, details in EXCEPTION_PATTERNS.items():
+            doc_title_match = pattern_match(doc_title.lower(), pattern)
+            allowed_for_match = pattern_match_list(actor_email, details.get("allowed_for"))
+            allowed_for_all_match = details.get("allowed_for") == {"all"}
+
+            if (doc_title_match and
+                    (allowed_for_match or allowed_for_all_match) and
+                    current_time < details.get("allowed_until")):
+                return False
+            # No exceptions match.
+            # Return the event summary (which is True) to alert & use in title.
+            return {
+                "actor": actor_email,
+                "doc_title": doc_title,
+                "target_user": target_user,
+            }
+    return False
+
+
+def rule(event):
+    application_name = deep_get(event, 'id', 'applicationName')
+    events = event.get("events")
+    actor_email = deep_get(event, 'actor', 'email', default="EMAIL_UNKNOWN")
+
+    if (application_name == "drive" and events and
+            "acl_change" in set(e["type"] for e in events)):
+        # If any of the events in this record are a dangerous file share, alert:
+        return any(
+            _check_acl_change_event(actor_email, acl_change_event)
+            for acl_change_event in events)
+    return False
+
+
+def title(event):
+    events = event.get("events", [])
+    actor_email = deep_get(event, 'actor', 'email', default='EMAIL_UNKNOWN')
+    matching_events = [
+        _check_acl_change_event(actor_email, acl_change_event)
+        for acl_change_event in events
+        if _check_acl_change_event(actor_email, acl_change_event)
+    ]
+
+    if matching_events:
+        len_events = len(matching_events)
+        first_event = matching_events[0]
+        actor = first_event.get("actor", "ACTOR_UNKNOWN")
+        doc_title = first_event.get("doc_title", "DOC_TITLE_UNKNOWN")
+        target_user = first_event.get("target_user", "USER_UNKNOWN")
+        if len(matching_events) > 1:
+            return f'Multiple dangerous shares ({len_events}) by [{actor}], including "{doc_title}" to {target_user}'
+        return f'Dangerous file share by [{actor}]: "{doc_title}" to {target_user}'
+    # How could there not be if title() gets called?
+    return "ERROR - No matching events, but DangerousShares still fired?"
```

Comment:
You could also raise an exception here and this would be treated as an erroring rule.

Reply:
Do we have a generic PantherRuleError exception?

Reply:
We define a few exceptions related to using the KV table in `global_helpers/panther_oss_helpers.py`, but we don't have a generic one. Maybe we could add

```class PantherDetectionLogic(Exception):
    """Error returned when a Panther detection encounters an unexpected situation."""
```

Or something like that to either `panther_base_helpers` or `panther_oss_helpers`.